File: service/Services/AIService.cs
using Newtonsoft.Json;
using PaintLearning.Service.DTO;

namespace PaintLearning.Service.Services;

public class AIService : IAIService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;

    public AIService(HttpClient httpClient, IConfiguration configuration)
    {
        _httpClient = httpClient;
        _configuration = configuration;
    }

    public async Task<AnalysisResultDTO> AnalyzeArtworkAsync(string imageBase64, string language = "zh")
    {
        return await CreateMockAnalysisAsync();
    }

    public async Task<TutorialDTO> GenerateTutorialAsync(AnalysisResultDTO analysis, string language = "zh")
    {
        return await CreateMockTutorialAsync();
    }

    public async Task<FeedbackDTO> ComparePracticeAsync(string originalImageBase64, string practiceImageBase64, string language = "zh")
    {
        return await CreateMockFeedbackAsync();
    }

    private async Task<AnalysisResultDTO> CreateMockAnalysisAsync()
    {
        await Task.Delay(500);
        
        return new AnalysisResultDTO
        {
            Id = Guid.NewGuid().ToString(),
            Composition = new CompositionAnalysisDTO
            {
                LayoutType = "三分法",
                FocalPoint = "画面中心偏右",
                SpatialHierarchy = "前景、中景、背景层次分明",
                NegativeSpace = "留白恰到好处",
                Details = "整体构图稳定，视觉引导自然"
            },
            Color = new ColorAnalysisDTO
            {
                PrimaryColors = new List<string> { "#FF6B6B", "#4ECDC4", "#45B7D1" },
                SchemeType = "互补色搭配",
                Temperature = "偏暖",
                Saturation = "中等偏高",
                Emotion = "温暖、活力",
                Details = "色彩搭配和谐，富有层次感"
            },
            Technique = new TechniqueAnalysisDTO
            {
                Medium = "油画",
                BrushStroke = "明显的笔触纹理",
                Texture = "厚重的颜料堆叠",
                SpecialTechniques = new List<string> { "厚涂法", "干笔技法" },
                Details = "技法熟练，表现力强"
            },
            Style = new StyleAnalysisDTO
            {
                Genre = "印象派",
                ArtistReference = "莫奈风格",
                Characteristics = new List<string> { "色彩鲜艳", "光影变化", "笔触灵动" },
                Details = "具有典型的印象派特征"
            },
            LightShadow = new LightShadowAnalysisDTO
            {
                LightDirection = "右上方",
                Contrast = "中等对比",
                Volume = "立体感强",
                Details = "光影处理得当，增强了画面深度"
            },
            OverallSummary = "这是一幅优秀的印象派风格油画作品，构图合理，色彩搭配和谐，技法娴熟。"
        };
    }

    private async Task<TutorialDTO> CreateMockTutorialAsync()
    {
        await Task.Delay(500);
        
        return new TutorialDTO
        {
            Id = Guid.NewGuid().ToString(),
            Steps = new List<TutorialStepDTO>
            {
                new TutorialStepDTO
                {
                    Order = 1,
                    Title = "准备阶段",
                    Description = "准备画布、颜料和画笔等工具材料",
                    KeyPoints = new List<string> { "选择合适的画布尺寸", "准备基本颜色" }
                },
                new TutorialStepDTO
                {
                    Order = 2,
                    Title = "起稿",
                    Description = "用铅笔勾勒大致轮廓",
                    KeyPoints = new List<string> { "注意构图比例", "确定主体位置" }
                },
                new TutorialStepDTO
                {
                    Order = 3,
                    Title = "铺大色块",
                    Description = "铺设背景和主体的基本色调",
                    KeyPoints = new List<string> { "从大面积开始", "注意色彩层次" }
                },
                new TutorialStepDTO
                {
                    Order = 4,
                    Title = "深入刻画",
                    Description = "细化细节和光影",
                    KeyPoints = new List<string> { "注意明暗对比", "保持笔触方向" }
                },
                new TutorialStepDTO
                {
                    Order = 5,
                    Title = "调整统一",
                    Description = "整体调整画面效果",
                    KeyPoints = new List<string> { "检查整体色调", "调整细节" }
                }
            },
            Materials = new List<string> { "油画布", "油画颜料", "画笔", "调色板" },
            Tips = new List<string> { "保持调色板清洁", "注意颜料厚度", "多观察实物" }
        };
    }

    private async Task<FeedbackDTO> CreateMockFeedbackAsync()
    {
        await Task.Delay(500);
        
        return new FeedbackDTO
        {
            Id = Guid.NewGuid().ToString(),
            Strengths = new List<string> { "色彩把握较好", "构图合理", "笔触流畅" },
            Weaknesses = new List<string> { "明暗对比不够强烈", "细节处理有待提高" },
            Suggestions = new List<SuggestionDTO>
            {
                new SuggestionDTO
                {
                    Area = "光影",
                    Description = "加强明暗对比，增强立体感",
                    Priority = "high",
                    IsResolved = false
                },
                new SuggestionDTO
                {
                    Area = "细节",
                    Description = "注意细节刻画，提升画面精致度",
                    Priority = "medium",
                    IsResolved = false
                }
            },
            OverallRating = 4
        };
    }
}

File: src/services/AIService.ts
import { AnalysisResult, Tutorial, Feedback } from '../types';

export interface AIService {
  analyzeArtwork(imageBase64: string, language: string): Promise<AnalysisResult>;
  generateTutorial(analysis: AnalysisResult, language: string): Promise<Tutorial>;
  comparePractice(originalImageBase64: string, practiceImageBase64: string, language: string): Promise<Feedback>;
}

export abstract class BaseAdapter implements AIService {
  protected apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  abstract analyzeArtwork(imageBase64: string, language: string): Promise<AnalysisResult>;
  abstract generateTutorial(analysis: AnalysisResult, language: string): Promise<Tutorial>;
  abstract comparePractice(originalImageBase64: string, practiceImageBase64: string, language: string): Promise<Feedback>;

  protected getLanguagePrompt(language: string): string {
    const langMap: Record<string, string> = {
      'zh': '中文',
      'en': 'English',
      'ja': '日本語',
      'ko': '한국어',
    };
    return langMap[language] || '中文';
  }
}

export class OpenAIAdapter extends BaseAdapter {
  async analyzeArtwork(imageBase64: string, language: string): Promise<AnalysisResult> {
    const prompt = `你是一位专业的艺术评论家。请从以下维度分析这幅画作，并以JSON格式返回结果，语言使用${this.getLanguagePrompt(language)}：

{
  "composition": {
    "layoutType": "布局类型（如三分法、黄金分割等）",
    "focalPoint": "视觉重心",
    "spatialHierarchy": "空间层次",
    "negativeSpace": "正负空间运用",
    "details": "详细分析"
  },
  "color": {
    "primaryColors": ["主色调1", "主色调2"],
    "schemeType": "配色方案类型",
    "temperature": "色彩冷暖",
    "saturation": "饱和度分析",
    "emotion": "色彩情感",
    "details": "详细分析"
  },
  "technique": {
    "medium": "绘画媒介（油画/水彩/素描等）",
    "brushStroke": "笔触特征",
    "texture": "肌理效果",
    "specialTechniques": ["特殊技法1", "特殊技法2"],
    "details": "详细分析"
  },
  "style": {
    "genre": "艺术流派",
    "artistReference": "参考艺术家",
    "characteristics": ["风格特征1", "风格特征2"],
    "details": "详细分析"
  },
  "lightShadow": {
    "lightDirection": "光源方向",
    "contrast": "明暗对比",
    "volume": "体积感塑造",
    "details": "详细分析"
  },
  "overallSummary": "总体评价"
}`;

    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: 'gpt-4o',
        messages: [
          {
            role: 'user',
            content: [
              { type: 'text', text: prompt },
              { type: 'image_url', image_url: { url: `data:image/jpeg;base64,${imageBase64}` } }
            ]
          }
        ],
        max_tokens: 2000,
        response_format: { type: 'json_object' }
      }),
    });

    if (!response.ok) {
      throw new Error('OpenAI API request failed');
    }

    const data = await response.json();
    return JSON.parse(data.choices[0].message.content);
  }

  async generateTutorial(analysis: AnalysisResult, language: string): Promise<Tutorial> {
    const prompt = `基于以下画作分析，生成一份分步骤绘画教程，语言使用${this.getLanguagePrompt(language)}，返回JSON格式：

分析结果：${JSON.stringify(analysis)}

{
  "materials": ["材料1", "材料2"],
  "tips": ["提示1", "提示2"],
  "steps": [
    {
      "order": 1,
      "title": "准备阶段",
      "description": "准备工作描述",
      "keyPoints": ["要点1", "要点2"],
      "regionHint": "画面区域提示"
    },
    {
      "order": 2,
      "title": "起稿",
      "description": "起稿步骤描述",
      "keyPoints": ["要点1", "要点2"]
    },
    {
      "order": 3,
      "title": "铺大色块",
      "description": "铺色步骤描述",
      "keyPoints": ["要点1", "要点2"]
    },
    {
      "order": 4,
      "title": "深入刻画",
      "description": "深入步骤描述",
      "keyPoints": ["要点1", "要点2"]
    },
    {
      "order": 5,
      "title": "调整统一",
      "description": "调整步骤描述",
      "keyPoints": ["要点1", "要点2"]
    }
  ]
}`;

    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: 'gpt-4o',
        messages: [{ role: 'user', content: prompt }],
        max_tokens: 2000,
        response_format: { type: 'json_object' }
      }),
    });

    if (!response.ok) {
      throw new Error('OpenAI API request failed');
    }

    const data = await response.json();
    return JSON.parse(data.choices[0].message.content);
  }

  async comparePractice(originalImageBase64: string, practiceImageBase64: string, language: string): Promise<Feedback> {
    const prompt = `请对比原画和练习作品，给出详细的反馈，语言使用${this.getLanguagePrompt(language)}，返回JSON格式：

{
  "strengths": ["优点1", "优点2"],
  "weaknesses": ["不足1", "不足2"],
  "overallRating": 4,
  "suggestions": [
    {
      "area": "方面",
      "description": "建议描述",
      "priority": "high",
      "isResolved": false
    }
  ]
}`;

    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: 'gpt-4o',
        messages: [
          {
            role: 'user',
            content: [
              { type: 'text', text: prompt },
              { type: 'image_url', image_url: { url: `data:image/jpeg;base64,${originalImageBase64}` } },
              { type: 'image_url', image_url: { url: `data:image/jpeg;base64,${practiceImageBase64}` } }
            ]
          }
        ],
        max_tokens: 2000,
        response_format: { type: 'json_object' }
      }),
    });

    if (!response.ok) {
      throw new Error('OpenAI API request failed');
    }

    const data = await response.json();
    return JSON.parse(data.choices[0].message.content);
  }
}

export class ClaudeAdapter extends BaseAdapter {
  async analyzeArtwork(imageBase64: string, language: string): Promise<AnalysisResult> {
    throw new Error('Claude adapter not implemented yet');
  }

  async generateTutorial(analysis: AnalysisResult, language: string): Promise<Tutorial> {
    throw new Error('Claude adapter not implemented yet');
  }

  async comparePractice(originalImageBase64: string, practiceImageBase64: string, language: string): Promise<Feedback> {
    throw new Error('Claude adapter not implemented yet');
  }
}

export class GeminiAdapter extends BaseAdapter {
  async analyzeArtwork(imageBase64: string, language: string): Promise<AnalysisResult> {
    throw new Error('Gemini adapter not implemented yet');
  }

  async generateTutorial(analysis: AnalysisResult, language: string): Promise<Tutorial> {
    throw new Error('Gemini adapter not implemented yet');
  }

  async comparePractice(originalImageBase64: string, practiceImageBase64: string, language: string): Promise<Feedback> {
    throw new Error('Gemini adapter not implemented yet');
  }
}

export class AIServiceFactory {
  static create(model: 'openai' | 'claude' | 'gemini', apiKey: string): AIService {
    switch (model) {
      case 'openai':
        return new OpenAIAdapter(apiKey);
      case 'claude':
        return new ClaudeAdapter(apiKey);
      case 'gemini':
        return new GeminiAdapter(apiKey);
      default:
        throw new Error(`Unknown model: ${model}`);
    }
  }
}
